File: javascript_version/graphs/longest_path/longest_path.js
// longest path
// Write a function, longestPath, that takes in an adjacency list for a 
// directed acyclic graph. The function should return the length of the 
// longest path within the graph.
//  A path may start and end at any two nodes. The length of a path is considered 
//  the number of edges in the path, not the number of nodes.


const longestPath = (graph) => {
    // todo

    let maxPath = 0;
    let hash = {};
    for (let node in graph) {
        if (graph[node].length === 0) {
            hash[node] = 0;
        }
    }
    for (let node in graph) {
        maxPath = Math.max(hasPath(graph, hash, node), maxPath);
    }
    return maxPath;
};

const hasPath = (graph, hash, node) => {
    if (node in hash) { return hash[node]; }
    let size = 0;
    for (let currNode of graph[node]) {

        let curr = hasPath(graph, hash, currNode);
        if (curr > size) { size = curr; }

    }
    hash[node] = size + 1;
    return hash[node];
}



let graph = {
    a: ['c', 'b'],
    b: ['c'],
    c: []
};

console.log(longestPath(graph)); // -> 2
graph = {
    a: ['c', 'b'],
    b: ['c'],
    c: [],
    q: ['r'],
    r: ['s', 'u', 't'],
    s: ['t'],
    t: ['u'],
    u: []
};

console.log(longestPath(graph)); // -> 4
graph = {
    h: ['i', 'j', 'k'],
    g: ['h'],
    i: [],
    j: [],
    k: [],
    x: ['y'],
    y: []
};

console.log(longestPath(graph)); // -> 2
graph = {
    a: ['b'],
    b: ['c'],
    c: [],
    e: ['f'],
    f: ['g'],
    g: ['h'],
    h: []
};

console.log(longestPath(graph)); // -> 3

File: java_version/graphs/longest_path/longest_path.java
package graphs.longest_path;

import java.util.*;

// longest path
// Write a function, longestPath, that takes in an adjacency list for a 
// directed acyclic graph. The function should return the length of the 
// longest path within the graph.
//  A path may start and end at any two nodes. The length of a path is considered 
//  the number of edges in the path, not the number of nodes.

public class longest_path {

    public static int longestPath(Map<Character, List<Character>> graph) {

        int maxPath = 0;


        return maxPath;
    }

    public static int hasPath(Map<Character, List<Character>> graph, char node, HashMap<Character, Integer> visited) {


    

        return visited.get(node);
    }

    public static void main(String[] args) {
        Map<Character, List<Character>> graph = Map.of(
                'a', List.of('c', 'b'),
                'b', List.of('c'),
                'c', List.of());

        System.out.println(longestPath(graph)); // -> 2

        graph = Map.of(
                'a', List.of('b', 'c'),
                'b', List.of('c'),
                'c', List.of(),
                'q', List.of('r'),
                'r', List.of('s', 'u', 't'),
                's', List.of('t'),
                't', List.of('u'),
                'u', List.of());
        System.out.println(longestPath(graph)); // -> 4

        graph = Map.of(
                'h', List.of('i', 'j', 'k'),
                'g', List.of('h'),
                'i', List.of(),
                'j', List.of(),
                'k', List.of(),
                'x', List.of('y'),
                'y', List.of());

        System.out.println(longestPath(graph));// -> 2

        graph = Map.of(
                'a', List.of('b'),
                'b', List.of('c'),
                'c', List.of(),
                'e', List.of('f'),
                'f', List.of('g'),
                'g', List.of('h'),
                'h', List.of());

        System.out.println(longestPath(graph)); // -> 3

    }
}